dotnet/src/SemanticKernel/Reliability/RetryWithDelay.cs

```diff
@@ -0,0 +1,57 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.SemanticKernel.AI;
+using Microsoft.SemanticKernel.Diagnostics;
+
+namespace Microsoft.SemanticKernel.Reliability;
+
+/// <summary>
+/// A retry mechanism that retries once after a delay, using RetryAfter header if available.
+/// </summary>
+public class RetryWithDelay : IRetryMechanism
+{
+    private int RetryCount { get; } = 2;
+    private TimeSpan RetryDelay { get; } = TimeSpan.FromSeconds(2);
+
+    public RetryWithDelay(int? retryCount = null, TimeSpan? retryDelay = null)
+    {
+        this.RetryCount = retryCount ?? this.RetryCount;
+        this.RetryDelay = retryDelay ?? this.RetryDelay;
+    }
+
+    public async Task ExecuteWithRetryAsync(Func<Task> action, ILogger log, CancellationToken cancellationToken = default)
+    {
+        for (var retryCount = 0; retryCount < this.RetryCount; retryCount++)
+        {
+            try
+            {
+                if (!cancellationToken.IsCancellationRequested)
+                {
+                    await action();
+                }
+
+                break;
+            }
+            catch (AIException ex) when ((ex.ErrorCode is AIException.ErrorCodes.Throttling or
```

Comment:
what about other transient errors, e.g. timeout and service unavailable?

Reply:
Addressed! ServiceError is added -- but we might want to break this up a bit...